### PR TITLE
fix query to filter on effective location

### PR DESCRIPTION
**Pull request by rhundal7**
- PR for issue https://github.com/TAMULib/fw-registry/issues/378
- Earlier I was setting the locationNames array to either get populated from the front end or set it to empty array and based my logic on that which was overcomplicated. I now changed it slightly. If the user selects all (more than allowed selections), in the frontend, it de-selects all selections, shows message to user. If the user still submits the form, it passes an empty array to backend, processes it (the scenario of everything) same like the scenario of nothing selected and returns the respective results. If the user selects some options, it passes that populated array to the backend and the respective where clause gets executed giving the result.

### CODE REVIEW

File: books-call-number/nodes/js/buildQuery.js [hunk 1]
```diff
@@ -1,40 +1,61 @@
-if (logLevel === 'DEBUG') {
```
Comment (rmathew1011) [inline books-call-number/nodes/js/buildQuery.js:1]: Could you have this if condition back.

File: books-call-number/nodes/js/buildQuery.js [hunk 2]
```diff
@@ -1,40 +1,61 @@
...
+
+
+if (totalCharacterCount > maxAllowedCharacters) {
+   throw new Error("Selection exceeds the maximum allowed character count.");
+}
+execution.setVariable('validatedLocationNames', locationNames);
+
 if (startRange) {
-  where = '\n\t\tUPPER(ie.effective_call_number) >= UPPER(\'' + startRange + '\')';
+where = '\n\t\tUPPER(ie.effective_call_number) >= UPPER(\'' + startRange + '\')';
```
Comment (rmathew1011) [inline books-call-number/nodes/js/buildQuery.js:27]: Check for the indentation here

File: books-call-number/nodes/js/buildQuery.js [hunk 3]
```diff
@@ -1,40 +1,61 @@
...
 }
 
 if (endRange) {
-  where += '\n\t\tAND UPPER(ie.effective_call_number) <= RPAD(UPPER(\'' + endRange + '\'), max_len, \'ÿ\')';
+where += '\n\t\tAND UPPER(ie.effective_call_number) <= RPAD(UPPER(\'' + endRange + '\'), max_len, \'ÿ\')';
```
Comment (rmathew1011) [inline books-call-number/nodes/js/buildQuery.js:31]: could you add extra space  - code is within if block.

File: books-call-number/nodes/js/buildQuery.js [hunk 4]
```diff
@@ -1,40 +1,61 @@
...
 }
 
 where += '\n\t\tAND ie.status_name = \'Checked out\'';
 
+if (locationNameArray.length > 0 && locationNameArray.length < totalCharacterCount) {
+ print(locationNameArray);
```
Comment (rmathew1011) [inline books-call-number/nodes/js/buildQuery.js:37]: could you add extra space - code is within if block.

File: books-call-number/nodes/js/buildQuery.js [hunk 5]
```diff
@@ -1,40 +1,61 @@
...
-  '\nSELECT ie.effective_call_number' +
-  '\n\tFROM folio_reporting.item_ext ie' +
-  '\n\tCROSS JOIN MaxLength' +
-  '\nWHERE ' + where +
-  '\nORDER BY ie.effective_call_number';
-
-if (logLevel === 'DEBUG') {
-  print('\nbooksCallNumberQuery = ' + booksCallNumberQuery);
-}
+'\n\n' + cte +
```
Comment (rmathew1011) [inline books-call-number/nodes/js/buildQuery.js:47]: could you bring in the  if condition.

File: books-call-number/nodes/js/buildQuery.js [hunk 6]
```diff
@@ -1,40 +1,61 @@
...
+'\n\tie.effective_location_name AS item_effective_location' +
+'\n\tFROM folio_reporting.item_ext ie' +
+'\n\tCROSS JOIN MaxLength' +
+'\nWHERE ' + where +
+'\nORDER BY ie.effective_call_number, item_effective_location';
+
+print(booksCallNumberQuery);
 
 var queryWrapper = {
-  sql: booksCallNumberQuery,
```
Comment (rmathew1011) [inline books-call-number/nodes/js/buildQuery.js:37]: could you add extra space - code is within a block.

**Review by rmathew1011 — CHANGES_REQUESTED** [review]

File: books-call-number/nodes/js/buildQuery.js [hunk 7]
```diff
@@ -1,40 +1,61 @@
...
+sql: booksCallNumberQuery,
 };
 
-execution.setVariableLocal('booksCallNumberQuery', S(JSON.stringify(queryWrapper)));
+execution.setVariableLocal('booksCallNumberQuery', S(JSON.stringify(queryWrapper)));
```
Comment (Dbreck-TAMU) [inline books-call-number/nodes/js/buildQuery.js:61]: Please add EoF Line. 
```suggestion
execution.setVariableLocal('booksCallNumberQuery', S(JSON.stringify(queryWrapper)));

```

File: books-call-number/nodes/js/buildQuery.js [hunk 8]
```diff
@@ -1,40 +1,61 @@
...
+
+for (var i = 0; i < locationNames.length; i++) {
+  var locationNameVar = locationNames[i];
+  totalCharacterCount += locationNameVar.length;
+}
+
+
+if (totalCharacterCount > maxAllowedCharacters) {
+   throw new Error("Selection exceeds the maximum allowed character count.");
+}
```
Comment (Dbreck-TAMU) [inline books-call-number/nodes/js/buildQuery.js:23]: Code suggestion for clarity: 

const locationNames = execution.getVariable('locationName') || [];
const MAX_ALLOWED_CHARACTERS = 3000;

if (MAX_ALLOWED_CHARACTERS < locationNames.length) {
  throw new Error("Selection exceeds the maximum allowed character count.");
}

var locationNameArray = JSON.parse(locationName);

Comment (rhundal7) [reply books-call-number/nodes/js/buildQuery.js:23]: I will most likely be removing this block 
```
if (totalCharacterCount > maxAllowedCharacters) {
   throw new Error("Selection exceeds the maximum allowed character count.");
}
```

File: books-call-number/nodes/js/buildQuery.js [hunk 9]
```diff
@@ -2,10 +2,22 @@ if (logLevel === 'DEBUG') {
   print('\nlogLevel = ' + logLevel + '\n');
   print('\ncall number start range = ' + startRange + '\n');
   print('\ncall number end range = ' + endRange + '\n');
+  print('locationName = ' + locationName + '\n');
 }
-
 var where = 'TRUE';
 
+const locationNames = execution.getVariable('locationName') || [];
```
Comment (wwelling) [inline books-call-number/nodes/js/buildQuery.js:9]:
```suggestion

```

File: books-call-number/nodes/js/buildQuery.js [hunk 10]
```diff
@@ -2,10 +2,22 @@ if (logLevel === 'DEBUG') {
...
+const maxAllowedCharacters = 3000;
+let totalCharacterCount = 0;
```
Comment (wwelling) [inline books-call-number/nodes/js/buildQuery.js:11]:
```suggestion
let totalCharacterCount = locationName.length;
```

File: books-call-number/nodes/js/buildQuery.js [hunk 11]
```diff
@@ -2,10 +2,22 @@ if (logLevel === 'DEBUG') {
...
+
+var locationNameArray = JSON.parse(locationName);
+
+for (var i = 0; i < locationNames.length; i++) {
```
Comment (wwelling) [inline books-call-number/nodes/js/buildQuery.js:15]: As locationNames is the serialized string on the context and it is what can exceed the history table max column for variable value, iterating over the string counting characters is not necessary.

File: books-call-number/nodes/js/buildQuery.js [hunk 12]
```diff
@@ -2,10 +2,22 @@ if (logLevel === 'DEBUG') {
   print('\nlogLevel = ' + logLevel + '\n');
   print('\ncall number start range = ' + startRange + '\n');
   print('\ncall number end range = ' + endRange + '\n');
+  print('locationName = ' + locationName + '\n');
```
Comment (wwelling) [inline books-call-number/nodes/js/buildQuery.js:5]:
```suggestion
  print('\nlocationName = ' + locationName + '\n');
```

**Review by wwelling — CHANGES_REQUESTED** [review]
No need to get variables using the execution context, they are already available in the JavaScript scope.

File: books-call-number/nodes/js/buildQuery.js [hunk 13]
```diff
@@ -16,22 +28,26 @@ if (endRange) {
 
 where += '\n\t\tAND ie.status_name = \'Checked out\'';
 
+if (locationNameArray.length > 0 && locationNameArray.length < totalCharacterCount) {
```
Comment (wwelling) [inline books-call-number/nodes/js/buildQuery.js:31]: locationNameArray.length < totalCharacterCount is always going to be less than the total number of characters summing the location names.
```suggestion
if (locationNameArray.length > 0 && totalCharacterCount  <= maxAllowedCharacters ) {
```

File: books-call-number/nodes/js/buildQuery.js [hunk 14]
```diff
@@ -2,10 +2,22 @@ if (logLevel === 'DEBUG') {
   print('\nlogLevel = ' + logLevel + '\n');
   print('\ncall number start range = ' + startRange + '\n');
   print('\ncall number end range = ' + endRange + '\n');
+  print('locationName = ' + locationName + '\n');
 }
-
 var where = 'TRUE';
 
+const locationNames = execution.getVariable('locationName') || [];
+const maxAllowedCharacters = 3000;
```
Comment (wwelling) [inline books-call-number/nodes/js/buildQuery.js:10]:
```suggestion
```

File: books-call-number/nodes/js/buildQuery.js (hunk 10 above)
Comment (wwelling) [inline books-call-number/nodes/js/buildQuery.js:11]:
```suggestion
```

File: books-call-number/nodes/js/buildQuery.js [hunk 15]
```diff
@@ -2,10 +2,15 @@ if (logLevel === 'DEBUG') {
   print('\nlogLevel = ' + logLevel + '\n');
   print('\ncall number start range = ' + startRange + '\n');
   print('\ncall number end range = ' + endRange + '\n');
+  print('locationName = ' + locationName + '\n');
 }
-
 var where = 'TRUE';
 
+const locationNames = execution.getVariable('locationName');
```
Comment (wwelling) [inline books-call-number/nodes/js/buildQuery.js:9]: Can this be removed and just use the variable `locationName` that is already on the scope?

File: books-call-number/nodes/js/buildQuery.js [hunk 16]
```diff
@@ -2,10 +2,12 @@ if (logLevel === 'DEBUG') {
   print('\nlogLevel = ' + logLevel + '\n');
   print('\ncall number start range = ' + startRange + '\n');
   print('\ncall number end range = ' + endRange + '\n');
+  print('\nlocationName = ' + locationName + '\n');
 }
-
 var where = 'TRUE';
```
Comment (kaladay) [inline books-call-number/nodes/js/buildQuery.js:7]:
```suggestion

var where = 'TRUE';
```
Add back missing new line.

File: books-call-number/nodes/js/buildQuery.js [hunk 17]
```diff
@@ -2,10 +2,12 @@ if (logLevel === 'DEBUG') {
...
 
+var locationNameArray = JSON.parse(locationName);
+
 if (startRange) {
```
Comment (kaladay) [inline books-call-number/nodes/js/buildQuery.js:11]:
```suggestion
if (!!startRange) {
```

Being that this is JavaScript, we do not want to rely of the boolean check on the variable.
Instead, use a double exclamation, like `!!`.

Please do the same for the `if (endRange) {` below.

File: books-call-number/nodes/js/buildQuery.js [hunk 18]
```diff
@@ -16,18 +18,23 @@ if (endRange) {
...
 where += '\n\t\tAND ie.status_name = \'Checked out\'';
 
+if (locationNameArray.length > 0) { 
+  where += "\n\tAND ie.effective_location_name IN ('"+locationNameArray.join("', '")+"')";
+}
+
 var cte = 'WITH MaxLength AS (' +
 '\n\tSELECT MAX(LENGTH(ie.effective_call_number)) AS max_len' +
 '\n\tFROM folio_reporting.item_ext ie' +
 ')';
```
Comment (kaladay) [inline books-call-number/nodes/js/buildQuery.js:28]:
```suggestion
  '\n\tSELECT MAX(LENGTH(ie.effective_call_number)) AS max_len' +
  '\n\tFROM folio_reporting.item_ext ie' +
  ')';
```

File: books-call-number/nodes/js/buildQuery.js [hunk 19]
```diff
@@ -16,18 +18,23 @@ if (endRange) {
 
 where += '\n\t\tAND ie.status_name = \'Checked out\'';
 
+if (locationNameArray.length > 0) { 
+  where += "\n\tAND ie.effective_location_name IN ('"+locationNameArray.join("', '")+"')";
```
Comment (kaladay) [inline books-call-number/nodes/js/buildQuery.js:22]:
```suggestion
  where += "\n\tAND ie.effective_location_name IN ('" + locationNameArray.join("', '") + "')";
```

**Review by kaladay — CHANGES_REQUESTED** [review]

**Review by kaladay — APPROVED** [review]

**Review by wwelling — APPROVED** [review]

**Review by rmathew1011 — APPROVED** [review]

**Pull request merged**